File: internal/api/attendance.go
package api

import (
	"net/http"
	"strings"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"

	"github.com/school-erp/erp/internal/httpx"
)

type attendanceRow struct {
	ID          string  `json:"id"`
	StudentID   string  `json:"student_id"`
	StudentName string  `json:"student_name"`
	AdmissionNo string  `json:"admission_no"`
	SectionID   string  `json:"section_id"`
	OnDate      string  `json:"on_date"`
	Status      string  `json:"status"`
	MinutesLate *int32  `json:"minutes_late,omitempty"`
	Remarks     *string `json:"remarks,omitempty"`
}

func (s *Server) listAttendance(w http.ResponseWriter, r *http.Request) {
	q := r.URL.Query()
	on := q.Get("on_date")
	if on == "" {
		on = time.Now().Format(time.DateOnly)
	}

	// Without this a teacher reads the register for every section in the
	// school: RLS admits the rows because they all belong to one institution.
	res, err := s.resolveScope(r)
	if err != nil {
		httpx.Internal(w, r, err)
		return
	}
	args := []any{on, nullString(q.Get("section_id")), nullString(q.Get("student_id"))}
	scopePred, scopeArgs := res.AttendancePredicate("sa", len(args)+1)
	args = append(args, scopeArgs...)

	items, err := collect(s, r, `
		SELECT sa.id::text, sa.student_id::text,
		       concat_ws(' ', st.first_name, st.middle_name, st.last_name),
		       st.admission_no, sa.section_id::text,
		       to_char(sa.on_date,'YYYY-MM-DD'), sa.status, sa.minutes_late, sa.remarks
		  FROM student_attendance sa
		  JOIN students st ON st.id = sa.student_id
		 WHERE sa.on_date = $1::date
		   AND ($2::uuid IS NULL OR sa.section_id = $2)
		   AND ($3::uuid IS NULL OR sa.student_id = $3)
		   AND `+scopePred+`
		 ORDER BY st.admission_no`, args,
		func(rows pgx.Rows) (attendanceRow, error) {
			var v attendanceRow
			return v, rows.Scan(&v.ID, &v.StudentID, &v.StudentName, &v.AdmissionNo,
				&v.SectionID, &v.OnDate, &v.Status, &v.MinutesLate, &v.Remarks)
		})
	respond(w, r, items, err)
}

type markAttendanceRequest struct {
	SectionID string `json:"section_id"`
	OnDate    string `json:"on_date"`
	PeriodID  string `json:"period_id,omitempty"`
	Entries   []struct {
		StudentID   string  `json:"student_id"`
		Status      string  `json:"status"`
		MinutesLate *int32  `json:"minutes_late,omitempty"`
		Remarks     *string `json:"remarks,omitempty"`
	} `json:"entries"`
	/* Which channels to tell the family on, beyond the app.

	   The in-app alert always goes and is not a choice: it costs nothing, it
	   is the record the parent can go back to, and a school that has bought no
	   gateway at all must still be able to tell a family their child is not in
	   the building. SMS, WhatsApp and email cost money per message, so they
	   are the teacher's to tick — some schools send them for every absence,
	   some only when they have rung and got no answer. */
	NotifyChannels []string `json:"notify_channels,omitempty"`
	// The one case a school does want it silent: back-filling a register from
	// a fortnight ago, where texting every parent about an absence they
	// already know about is worse than useless.
	Silent bool `json:"silent,omitempty"`
}

// markAttendance upserts a whole section in one transaction.
//
// A class teacher marks 40 students as one action, so this must be atomic: a
// partial write would leave the register half-marked with no indication of
// where it stopped. The unique index on (student_id, on_date, period_id) makes
// the re-mark case an update rather than a duplicate row.
func (s *Server) markAttendance(w http.ResponseWriter, r *http.Request) {
	id := httpx.IdentityFrom(r.Context())

	var req markAttendanceRequest
	if !httpx.Decode(w, r, &req) {
		return
	}
	sectionID, err := uuid.Parse(req.SectionID)
	if err != nil {
		httpx.BadRequest(w, r, "section_id must be a uuid")
		return
	}
	if req.OnDate == "" {
		req.OnDate = time.Now().Format(time.DateOnly)
	}
	if _, err := time.Parse(time.DateOnly, req.OnDate); err != nil {
		httpx.BadRequest(w, r, "on_date must be YYYY-MM-DD")
		return
	}
	if len(req.Entries) == 0 {
		httpx.BadRequest(w, r, "entries must not be empty")
		return
	}
	// Exactly the set student_attendance_status_check allows. Anything else
	// would fail at COMMIT with a constraint violation the user cannot act on.
	valid := map[string]bool{"present": true, "absent": true, "late": true,
		"half_day": true, "leave": true, "holiday": true}
	for _, e := range req.Entries {
		if !valid[e.Status] {
			httpx.BadRequest(w, r, "invalid status: "+e.Status)
			return
		}
	}

	// Marking is limited to sections the caller teaches or is class teacher of,
	// unless they hold academics.attendance.write.any. Checking only the
	// permission let any teacher post a register for any class in the school.
	res, err := s.resolveScope(r)
	if err != nil {
		httpx.Internal(w, r, err)
		return
	}
	if !res.CanMarkSection(sectionID) {
		httpx.Forbidden(w, r, "academics.attendance.write for this section")
		return
	}

	var periodID *uuid.UUID
	if req.PeriodID != "" {
		p, err := uuid.Parse(req.PeriodID)
		if err != nil {
			httpx.BadRequest(w, r, "period_id must be a uuid")
			return
		}
		periodID = &p
	}

	// Uniqueness is enforced by two *partial* indexes, not one plain index:
	//
	//   student_attendance_daily  (student_id, on_date)             WHERE period_id IS NULL
	//   student_attendance_period (student_id, on_date, period_id)  WHERE period_id IS NOT NULL
	//
	// A conflict target must reproduce the index predicate to match, so
	// day-level and period-level marking need different ON CONFLICT clauses.
	// Omitting the WHERE gives "no unique or exclusion constraint matching the
	// ON CONFLICT specification" at runtime.
	const upsertTail = ` DO UPDATE
		   SET status         = EXCLUDED.status,
		       minutes_late   = EXCLUDED.minutes_late,
		       remarks        = EXCLUDED.remarks,
		       corrected_from = student_attendance.status,
		       corrected_by   = EXCLUDED.marked_by,
		       corrected_at   = now()
		 WHERE student_attendance.status IS DISTINCT FROM EXCLUDED.status`

	conflict := `ON CONFLICT (student_id, on_date) WHERE period_id IS NULL`
	if periodID != nil {
		conflict = `ON CONFLICT (student_id, on_date, period_id) WHERE period_id IS NOT NULL`
	}
	sql := `
		INSERT INTO student_attendance
		    (institution_id, student_id, section_id, on_date, period_id,
		     status, minutes_late, remarks, marked_by, marked_at)
		VALUES ($1,$2,$3,$4::date,$5,$6,$7,$8,$9, now())
		` + conflict + upsertTail

	channels := cleanChannels(req.NotifyChannels)
	written := 0
	told, queued := 0, 0
	var nowAbsent []uuid.UUID
	err = s.DB.InTenant(r.Context(), tenantScope(id), func(tx pgx.Tx) error {
		// Derive the tenant from the section rather than the caller: a platform
		// operator has no institution_id of their own, and the column is NOT NULL.
		var instID uuid.UUID
		if err := tx.QueryRow(r.Context(),
			`SELECT institution_id FROM sections WHERE id = $1`, sectionID).Scan(&instID); err != nil {
			return err
		}

		batch := &pgx.Batch{}
		for _, e := range req.Entries {
			studentID, err := uuid.Parse(e.StudentID)
			if err != nil {
				return err
			}
			batch.Queue(sql,
				instID, studentID, sectionID, req.OnDate, periodID,
				e.Status, e.MinutesLate, e.Remarks, id.UserID)
		}
		res := tx.SendBatch(r.Context(), batch)
		defer res.Close()
		/* WHICH CHILDREN ACTUALLY CHANGED, not which were submitted.

		   The upsert ends `WHERE status IS DISTINCT FROM EXCLUDED.status`, so
		   re-saving a register writes nothing — which is right, and it is also
		   the only thing standing between a teacher who presses Save twice and
		   a family that gets told twice. Notifying from req.Entries would text
		   every absent parent again on every save; notifying from the rows the
		   batch reports as written tells them once. */
		for i := range req.Entries {
			tag, err := res.Exec()
			if err != nil {
				return err
			}
			n := int(tag.RowsAffected())
			written += n
			if n > 0 && req.Entries[i].Status == "absent" {
				if sid, err := uuid.Parse(req.Entries[i].StudentID); err == nil {
					nowAbsent = append(nowAbsent, sid)
				}
			}
		}
		if err := res.Close(); err != nil {
			return err
		}
		/* Inside the same transaction as the register.

		   A notification about an absence the school then failed to record is
		   the worse of the two failures: the parent rings, and nobody at the
		   school can see what they are ringing about. */
		if !req.Silent && len(nowAbsent) > 0 {
			var qerr error
			told, queued, qerr = s.announceAbsences(
				r, tx, instID, nowAbsent, req.OnDate, channels)
			if qerr != nil {
				return qerr
			}
		}
		return nil
	})
	if err != nil {
		httpx.Internal(w, r, err)
		return
	}
	httpx.JSON(w, http.StatusOK, map[string]any{
		"section_id": req.SectionID,
		"on_date":    req.OnDate,
		"submitted":  len(req.Entries),
		"written":    written,
		// So the screen can say what actually happened rather than "Saved":
		// "31 marked, 3 absent, 5 parents told, 6 messages sent".
		"newly_absent":    len(nowAbsent),
		"parents_told":    told,
		"messages_queued": queued,
		"channels":        channels,
	})
}

// getMyStudent is the student/parent portal entry point: resolve the signed-in
// user to their own record without needing students.read.
func (s *Server) getMyStudent(w http.ResponseWriter, r *http.Request) {
	id := httpx.IdentityFrom(r.Context())

	out := map[string]any{}
	err := s.DB.InTenant(r.Context(), tenantScope(id), func(tx pgx.Tx) error {
		var (
			sid, admission, name string
			className, section   *string
		)
		err := tx.QueryRow(r.Context(), `
			SELECT st.id::text, st.admission_no,
			       concat_ws(' ', st.first_name, st.middle_name, st.last_name),
			       c.name, sec.name
			  FROM students st
			  LEFT JOIN LATERAL (
			      SELECT e.class_id, e.section_id FROM enrollments e
			       WHERE e.student_id = st.id ORDER BY e.enrolled_on DESC LIMIT 1
			  ) en ON true
			  LEFT JOIN classes c    ON c.id = en.class_id
			  LEFT JOIN sections sec ON sec.id = en.section_id
			 WHERE st.user_id = $1`, id.UserID).
			Scan(&sid, &admission, &name, &className, &section)
		if err != nil {
			return err
		}
		out["id"] = sid
		out["admission_no"] = admission
		out["full_name"] = name
		out["class_name"] = className
		out["section_name"] = section

		var present, total int
		if err := tx.QueryRow(r.Context(), `
			SELECT count(*) FILTER (WHERE status IN ('present','late')), count(*)
			  FROM student_attendance
			 WHERE student_id = $1 AND on_date >= date_trunc('month', CURRENT_DATE)`,
			sid).Scan(&present, &total); err != nil {
			return err
		}
		out["attendance_this_month"] = map[string]int{"present": present, "total": total}
		return nil
	})
	if err == pgx.ErrNoRows {
		httpx.NotFound(w, r)
		return
	}
	if err != nil {
		httpx.Internal(w, r, err)
		return
	}
	httpx.JSON(w, http.StatusOK, out)
}

/*
Telling a family their child is not in the building.

	THE WHOLE HOUSEHOLD, not the primary guardian. The manual absence run this
	replaces joined `sg.is_primary` and sent one SMS — so a mother who was not
	ticked as the primary contact was never told her child had not arrived, and
	the primary contact is a billing and consent flag that somebody in the
	office set years ago for reasons that had nothing to do with this.

	The child is told too, where they have an account. A teenager who is marked
	absent by mistake because the register was taken before they walked in is
	the person best placed to say so, and fastest.

	THE APP ALERT IS NOT OPTIONAL. It is free, it is the record the parent can
	go back to a week later, and it is the only channel that works for a school
	that has bought no gateway. The paid channels are the teacher's tick boxes.
*/
func (s *Server) announceAbsences(
	r *http.Request, tx pgx.Tx, inst uuid.UUID,
	students []uuid.UUID, onDate string, channels []string,
) (told, queued int, err error) {

	rows, err := tx.Query(r.Context(), `
		SELECT st.id,
		       concat_ws(' ', st.first_name, st.last_name),
		       to_char($2::date, 'DD Mon'),
		       p.user_id, p.phone, p.email::text
		  FROM students st
		  JOIN LATERAL (
		      SELECT g.user_id, g.phone, g.email::text AS email
		        FROM student_guardians sg
		        JOIN guardians g ON g.id = sg.guardian_id
		       WHERE sg.student_id = st.id
		      UNION ALL
		      -- The child's own account, where they have one.
		      SELECT u.id, u.phone, u.email::text FROM users u WHERE u.id = st.user_id
		  ) p ON true
		 WHERE st.id = ANY($1)`, students, onDate)
	if err != nil {
		return 0, 0, err
	}
	type msg struct {
		student      uuid.UUID
		name, date   string
		user         *uuid.UUID
		phone, email *string
	}
	var all []msg
	for rows.Next() {
		var m msg
		if err := rows.Scan(&m.student, &m.name, &m.date, &m.user, &m.phone, &m.email); err != nil {
			rows.Close()
			return 0, 0, err
		}
		all = append(all, m)
	}
	rows.Close()
	if err := rows.Err(); err != nil {
		return 0, 0, err
	}

	for _, m := range all {
		title := m.name + " was marked absent"
		body := m.name + " was marked absent on " + m.date +
			". If this is wrong, please tell the class teacher."

		if m.user != nil {
			sid := m.student
			if err := notify(r, tx, inst, *m.user, &sid, "attendance",
				title, body, "/portal/attendance", "student", &sid); err != nil {
				return told, queued, err
			}
			told++
		}
		for _, ch := range channels {
			to := ""
			if ch == "email" && m.email != nil {
				to = strings.TrimSpace(*m.email)
			} else if ch != "email" && m.phone != nil {
				to = strings.TrimSpace(*m.phone)
			}
			if to == "" {
				// No number, no address. Skipped rather than failed: the app
				// alert has already gone, and a register must not refuse to
				// save because one family has no mobile.
				continue
			}
			if _, err := s.QueueMessage(r.Context(), tx, inst, SendRequest{
				Channel:      ch,
				TemplateCode: "messaging.direct",
				Vars:         map[string]any{"text": body, "subject": title},
				Recipient:    to,
			}); err != nil {
				/* A gateway the school has not configured is theirs to fix and
				   is NOT a reason to fail the register. The teacher's actual
				   job here is recording who was in the room; losing that
				   because WhatsApp is misconfigured would be the product
				   choosing the wrong one of the two to protect. */
				continue
			}
			queued++
		}
	}
	return told, queued, nil
}

File: worker/src/routes/hr/attendance.ts
import type { Router } from '../../router'
import { enqueueMessageSends } from '../../services/messaging'
import { json } from '../../env'
import { HttpError, badRequest, created, forbidden, notFound, now, ok, readJSON, uuid, uuidParam } from '../../http'
import { can } from '../../identity'
import { addDays, fullName, isUUIDish, notImplemented, nz, str, todayIST } from '../admissions/util'

/* Port of the /attendance-workflow group (mod_academics.go, module 3) and
   the /operations dashboard (role_backoffice.go). */

const omitNull = <T extends object>(o: T): T => {
  for (const k of Object.keys(o) as (keyof T)[]) if (o[k] === null || o[k] === undefined) delete o[k]
  return o
}

/** scope.CanMarkSection: the class teacher of the section, or anyone who may mark any register. */
async function canMarkSection(c: { db: D1Database; id: import('../../identity').Identity }, sectionID: string): Promise<boolean> {
  if (c.id.platformAdmin || can(c.id, 'academics.attendance.write.any')) return true
  const row = await c.db.prepare(`SELECT 1 FROM sections WHERE id = ? AND class_teacher_id = ?`).bind(sectionID, c.id.userId).first()
  return !!row
}

/** requireOpenPeriod (period_close.go), kind = month: a month closed by hand, or inside a closed year. */
async function requireOpenMonth(db: D1Database, inst: string, onDate: string): Promise<void> {
  const row = await db.prepare(`
    SELECT EXISTS (SELECT 1 FROM period_closes WHERE institution_id = ? AND kind = 'month' AND period_key = ? AND reopened_at IS NULL) AS month_closed,
           (SELECT name FROM academic_years WHERE institution_id = ? AND closed_at IS NOT NULL AND ? BETWEEN starts_on AND ends_on ORDER BY starts_on DESC LIMIT 1) AS year_name`)
    .bind(inst, onDate.slice(0, 7), inst, onDate).first<{ month_closed: number; year_name: string | null }>()
  if (row?.month_closed) {
    const MONTHS = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December']
    throw new HttpError(409, `${MONTHS[Number(onDate.slice(5, 7)) - 1]} ${onDate.slice(0, 4)} is closed; ask the principal to reopen it`, { code: 'period_closed' })
  }
  if (row?.year_name) throw new HttpError(409, `The year ${row.year_name} is closed; ask the principal to reopen it`, { code: 'period_closed' })
}

export function registerAttendanceWorkflow(r: Router) {
  r.post('/attendance-workflow/corrections', 'academics.attendance.write', async (c) => {
    const req = await readJSON(c.req)
    const attID = str(req.attendance_id)
    if (!isUUIDish(attID)) throw badRequest('attendance_id must be a uuid')
    if (str(req.reason).trim() === '') throw badRequest('reason is required')
    const att = await c.db.prepare(`SELECT status, section_id FROM student_attendance WHERE id = ?`).bind(attID).first<{ status: string; section_id: string }>()
    if (!att) throw notFound()
    if (!(await canMarkSection(c, att.section_id))) throw forbidden('attendance correction for this section')
    const id = uuid()
    await c.db.prepare(`INSERT INTO attendance_corrections (id, institution_id, attendance_id, requested_by, from_status, to_status, reason, status, created_at) VALUES (?,?,?,?,?,?,?,'pending',?)`)
      .bind(id, c.id.institution!.id, attID, c.id.userId, att.status, str(req.to_status), str(req.reason), now()).run()
    return created({ id, status: 'pending' })
  })

  r.get('/attendance-workflow/corrections', 'academics.attendance.read', async (c) => {
    const status = nz(c.url.searchParams.get('status'))
    const rows = await c.db.prepare(`
      SELECT ac.id, ${fullName('st.first_name', 'st.last_name')} AS student_name, sa.on_date, ac.from_status, ac.to_status, ac.reason, u.full_name AS requested_by, ac.status
        FROM attendance_corrections ac JOIN student_attendance sa ON sa.id = ac.attendance_id JOIN students st ON st.id = sa.student_id LEFT JOIN users u ON u.id = ac.requested_by
       WHERE (? IS NULL OR ac.status = ?) ORDER BY ac.created_at DESC LIMIT 200`).bind(status, status).all<Record<string, unknown>>()
    return ok({ items: rows.results.map(omitNull) })
  })

  r.post('/attendance-workflow/corrections/{id}/decide', 'academics.attendance.write.any', async (c) => {
    const cid = uuidParam(c.params.id)
    const req = await readJSON(c.req)
    const decision = str(req.decision)
    if (decision !== 'approved' && decision !== 'rejected') throw badRequest('decision must be approved or rejected')
    const pending = await c.db.prepare(`SELECT ac.attendance_id, ac.to_status, ac.requested_by, sa.on_date, sa.institution_id, ${fullName('st.first_name', 'st.last_name')} AS child
        FROM attendance_corrections ac JOIN student_attendance sa ON sa.id = ac.attendance_id JOIN students st ON st.id = sa.student_id WHERE ac.id = ? AND ac.status = 'pending'`).bind(cid)
      .first<{ attendance_id: string; to_status: string; requested_by: string | null; on_date: string; institution_id: string; child: string }>()
    if (!pending) throw new HttpError(404, 'no pending correction with that id', { code: 'not_found' })
    if (decision === 'approved') await requireOpenMonth(c.db, pending.institution_id, pending.on_date)
    const t = now()
    const stmts = [c.db.prepare(`UPDATE attendance_corrections SET status = ?, decided_by = ?, decided_at = ? WHERE id = ? AND status = 'pending'`).bind(decision, c.id.userId, t, cid)]
    if (decision === 'approved') {
      stmts.push(c.db.prepare(`UPDATE student_attendance SET corrected_from = status, status = ?, corrected_by = ?, corrected_at = ? WHERE id = ?`).bind(pending.to_status, c.id.userId, t, pending.attendance_id))
    }
    // And tell whoever asked, unless they decided their own request.
    if (pending.requested_by && pending.requested_by !== c.id.userId) {
      const verb = decision === 'approved' ? 'approved' : 'not approved'
      const body = decision === 'approved' ? `The register for ${pending.on_date} has been amended.` : `The mark for ${pending.on_date} stands as it was. Raise it again if it is still wrong.`
      const dup = await c.db.prepare(`SELECT 1 FROM notifications WHERE user_id = ? AND kind = 'correction_decided' AND source_id = ? AND student_id IS NULL`).bind(pending.requested_by, cid).first()
      if (!dup) {
        stmts.push(c.db.prepare(`INSERT INTO notifications (id, institution_id, user_id, student_id, kind, title, body, link, source_kind, source_id, created_at) VALUES (?,?,?,NULL,'correction_decided',?,?,'/go/attendance_correction','attendance_correction',?,?)`)
          .bind(uuid(), c.id.institution!.id, pending.requested_by, `Your correction for ${pending.child} was ${verb}`, body, cid, t))
      }
    }
    const res = await c.db.batch(stmts)
    if ((res[0]?.meta.changes ?? 0) === 0) throw new HttpError(404, 'no pending correction with that id', { code: 'not_found' })
    return ok({ id: cid, status: decision })
  })

  r.post('/attendance-workflow/absence-alerts', 'comms.messages.send', async (c) => {
    // sendAbsenceAlerts: one attendance.absent SMS job per absent student's primary guardian.
    const onDate = c.url.searchParams.get('on_date') ?? ''
    const day = onDate !== '' ? onDate : todayIST()
    const targets = (await c.db.prepare(`SELECT g.user_id, trim(COALESCE(st.first_name,'') || ' ' || COALESCE(st.last_name,'')) AS student
        FROM student_attendance sa JOIN students st ON st.id = sa.student_id
        JOIN student_guardians sg ON sg.student_id = st.id AND sg.is_primary JOIN guardians g ON g.id = sg.guardian_id
       WHERE substr(sa.on_date,1,10) = ? AND sa.status = 'absent' AND g.user_id IS NOT NULL`).bind(day).all<{ user_id: string; student: string }>()).results
    await enqueueMessageSends(c.env, c.id.institution!.id, targets.map((t) => ({ channel: 'sms', template_key: 'attendance.absent', to_user_id: t.user_id,
      vars: { student: t.student, date: onDate } })))
    return json({ absent_students: targets.length, messages_queued: targets.length }, 202)
  })

  /* /operations/dashboard carries RequireAnyPermission over the four operations reads; the router
     has no any-of form, so it is registered on 'auth' and the handler checks the four itself. */
  r.get('/operations/dashboard', 'auth', async (c) => {
    if (!['operations.library.read', 'operations.transport.read', 'operations.hostel.read', 'operations.inventory.read'].some((p) => can(c.id, p))) throw forbidden()
    const today = todayIST()
    const k = await c.db.prepare(`
      SELECT (SELECT count(*) FROM library_titles) AS library_titles,
             (SELECT count(*) FROM library_loans WHERE returned_on IS NULL) AS loans_out,
             (SELECT count(*) FROM library_loans WHERE returned_on IS NULL AND due_on < ?) AS loans_overdue,
             (SELECT count(*) FROM vehicles WHERE status = 'active') AS vehicles,
             (SELECT count(*) FROM routes WHERE is_active = 1) AS routes,
             (SELECT count(*) FROM vehicles WHERE status='active' AND MIN(COALESCE(insurance_expiry,'9999-12-31'), COALESCE(fitness_expiry,'9999-12-31'), COALESCE(permit_expiry,'9999-12-31'), COALESCE(puc_expiry,'9999-12-31')) <= ?) AS vehicle_docs_expiring,
             (SELECT count(*) FROM transport_allocations) AS hostel_students`).bind(today, addDays(today, 30)).first()
    return ok(k)
  })
}
